torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -25,14 +27,20 @@ def foreach_all_gather(
     all_gather_copy_in_stream: torch.cuda.Stream,
     all_gather_stream: torch.cuda.Stream,
     device: torch.device,
-    dtype: torch.dtype,
 ) -> Optional[AllGatherResult]:
     world_size, rank = group.size(), group.rank()
     # - Copy in
     with torch.cuda.stream(all_gather_copy_in_stream):
         param_all_gather_inputs = [
             fsdp_param.all_gather_input for fsdp_param in fsdp_params
         ]
+        g = itertools.groupby(t.dtype for t in param_all_gather_inputs)
```

Comment:
Do we want to sort the inputs by dtype here so that the groups are contiguous? Otherwise, you'll have way more groups than there are dtypes.

Reply:
This is meant to be a fancy of way of checking that they are all the same, but I agree this is not so readable (taken from https://stackoverflow.com/questions/3844801/check-if-all-elements-in-a-list-are-identical).

Maybe we should just replace it with `all(t.dtype == param_all_gather_inputs[0].dtype for t in apram_all_gather_inputs)`, which is more obvious and not much slower.